test/linter/test-notes.js

```diff
@@ -0,0 +1,217 @@
+      });
+    }
+  }
+  for (let childNode of node.childNodes) {
+    testNode(childNode, browser, relPath, errors);
+  }
+};
+
+/**
+ * Test a string for valid HTML
+ *
+ * @param {string} string The string to test
+ * @param {string} browser The browser the notes belong to
+ * @param {string} relPath The identifier of the feature
+ * @param {HTMLError[]} errors The array to push errors to
+ * @returns {void}
+ */
+const validateHTML = (string, browser, relPath, errors) => {
+  const report = validator.validateString(string, {
+    rules: {
+      'attr-quotes': 'off',
+    },
+  });
+  if (report.valid) {
+    // If HTML is valid, ensure we're only using valid elements
+    let data = parser.parse(string);
+    testNode(data, browser, relPath, errors);
+  } else {
+    errors.push({
+      type: 'invalid',
+      messages: report.results
+        .map(x => x.messages)
+        .flat()
+        .map(x => x.message),
+      relPath,
+      browser,
+    });
+  }
+
+  if (string.includes('  ')) {
+    errors.push({
+      type: 'doublespace',
+      relPath,
+      browser,
+    });
+  }
+};
+
+/**
+ * Recursively check the notes in the data
+ *
+ * @param {string|string[]} notes The notes to test
+ * @param {string} browser The browser the notes belong to
+ * @param {string} relPath The identifier of the feature
+ * @param {HTMLError[]} errors The array to push errors to
+ * @returns {void}
+ */
+const checkNotes = (notes, browser, relPath, errors) => {
+  if (Array.isArray(notes)) {
+    for (let note of notes) {
+      checkNotes(note, browser, relPath, errors);
+    }
+  } else {
+    validateHTML(notes, browser, relPath, errors);
+  }
+};
+
+/**
+ * Process the data for notes errors
+ *
+ * @param {Identifier} data The data to test
+ * @param {HTMLError[]} errors The array to push errors to
+ * @param {string} [relPath] The identifier of the feature
+ * @returns {void}
+ */
+const processData = (data, errors, relPath) => {
+  for (const prop in data) {
+    if (prop === '__compat' && data[prop].support) {
+      let statement = data[prop].support;
+      for (const browser in statement) {
+        if (Array.isArray(statement[browser])) {
+          for (let s in statement[browser]) {
+            if (s.notes) checkNotes(s.notes, browser, relPath, errors);
+          }
+        } else {
+          if (statement[browser].notes)
+            checkNotes(statement[browser].notes, browser, relPath, errors);
+        }
+      }
+    }
+    const sub = data[prop];
+    if (typeof sub === 'object') {
+      processData(sub, errors, relPath ? `${relPath}.${prop}` : `${prop}`);
+    }
+  }
+};
+
+/**
+ * Test the data for notes errors
+ *
+ * @param {string} filename The file to test
+ * @returns {boolean} Whether the file had any errors
+ */
+const testNotes = filename => {
+  /** @type {Identifier} */
+  const data = require(filename);
+
+  /** @type {HTMLError[]} */
+  const errors = [];
+
+  processData(data, errors);
+
+  if (errors.length) {
```

Comment:
How about an early return to reduce the amount of indentation that follows?

```suggestion
  if (!errors.length) {
    return false;
  }
```